### PR TITLE
Update macOS Xcode tool checks do use Version class for comparisons

### PR DESCRIPTION
**Pull request by robinkunde**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Right now, Library/Homebrew/os/mac/xcode.rb uses string comparisons to determine if the version of a local tool is outdated. This will create false results if the two version strings are of different lengths, ie. "1000" < "800" will return true instead of false. Brew already has a class for proper comparison of semantic version strings and this PR updates xcode.rb to make use of it.

I have not added any tests since that would require somehow mocking the data returned by the system commands that xcode.rb uses to determine the versions of the installed tools, and I don't know how to do that at the moment.

### CODE REVIEW

**Comment by MikeMcQuaid** [comment]
> This will create false results if the two version strings are of different lengths, ie. "1000" < "800" will return true instead of false.

Can you provide an example within the usage of this class that currently breaks? Thanks!

**Comment by robinkunde** [comment]
I can't, and given the way version numbers have been increasing it _shouldn't_ be a problem until Xcode 10 becomes available. However, because the code is wrong I saw this as an appropriate thing to fix now, before we get blindsided.

For example, the length of the clang version string just changed. It's conceivable that Apple might put out version "800.0.42.2", and then "800.0.42.10" some time later. Using string comparison, "800.0.42.10" evaluate as smaller than "800.0.42.2"

**Comment by MikeMcQuaid** [comment]
@robinkunde I'm happy to edit files to get a reproduction but I'd like to be able to reproduce this issue before making changes here otherwise we're speculatively fixing something. Thanks!

**Comment by robinkunde** [comment]
@MikeMcQuaid oh, I see what you mean.

Edit Library/Homebrew/os/mac/xcode.rb. Hardcode uncached_version to always return "10.0" (line 94 or so). Then run "brew install git" for example. I now get this error:
Error: Your Xcode (10.0) is outdated.
Please update to Xcode 8.1 (or delete it).
Xcode can be updated from the App Store.

Now revert the change and bump the latest xcode version for your macOS release to "10.0" (line 19 or so). Run "brew install git" again and you get no complaint about Xcode being out of date.

**Comment by MikeMcQuaid** [comment]
@robinkunde Beautiful 😉! Nice spot. I think this code could be cleaned up perhaps with an `XcodeVersion` subclass that allows string comparisons which would avoid needing quite as many new `Version`s being created. Thoughts?

**Comment by robinkunde** [comment]
@MikeMcQuaid Is your concern the code repetition or the overhead of creating the objects? 
If it's the former, it could make sense to abstract that away. I thought about having the methods that return version strings right now return version objects directly, but I have no idea in how many different places those methods are called, and I didn't want to trigger any refactoring since there's poor test coverage right now.
If it's the latter, the overhead would just be moved into this subclass, unless we reimplement simple semantic version comparison by splitting the string on each dot and then compare each component as integers.

**Comment by MikeMcQuaid** [comment]
It was the former, yeh 👍

**Comment by robinkunde** [comment]
@MikeMcQuaid Ok, I gave it a go. Let me know what you think. I don't have a good grasp on idiomatic ruby yet, as I'm quite new to it.

File: Library/Homebrew/os/mac/xcode.rb [hunk 1]
```diff
@@ -225,16 +225,36 @@ def version
...
+          if XcodeVersion.is_higher_or_equal(MacOS.version, than: :mavericks.to_s)
             next unless File.exist?("#{MAVERICKS_PKG_PATH}/usr/bin/clang")
           end
           version = MacOS.pkgutil_info(id)[/version: (.+)$/, 1]
           return version if version
         end
       end
     end
+
+    module XcodeVersion
```
Comment (MikeMcQuaid) [inline Library/Homebrew/os/mac/xcode.rb]: Make this a class that inherits from `Version` instead and override `<=>`. Also, `Xcode::Version` is probably a preferable name. Thanks!

**Comment by robinkunde** [comment]
@MikeMcQuaid Sorry, it took me awhile to make time. I've amended the last commit with your suggestions. Let me know if that's what you were looking for. Thank you!

**Comment by MikeMcQuaid** [comment]
Much cleaner, thanks!

**Comment by MikeMcQuaid** [comment]
Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

**Pull request merged**